### ISSUE TITLE
Should implement a context menu widget

### ISSUE THREAD

**Issue by gspencergoog** [flutter/flutter#31955]
With the ability to right-click, we need to be able to allow developers to implement custom context menus.  On desktop platforms, these may need to be platform-specific context menus (so that they can render outside of the window), on web they will need to integrate with the context menu APIs there, and on Android they will need to be rendered by Flutter (one can right click on ChromeOS in an Android app, for instance. iOS doesn't support mouse, so it's not needed there). As a first pass on desktop platforms, it may initially be acceptable if the menu is rendered by Flutter (and hence doesn't render outside of the window), but long term probably not, so the API should definitely take that into account (and not take Widgets for the items, but rather some form of data object).

Before this, #31946 will probably need to be implemented, in order to provide the actions that the menu will invoke.

See also #30976 and #31412

**Comment by gspencergoog** [comment]
Actually, Flutter might be able to render the context menu itself on Windows, macOS, and Linux, since we can probably do this once we have multi-window support described in #30701, and still render outside of the window. This would allow a lot of flexibility in the visual presentation of the context menu (including looking just like the platform native context menu).

**Comment by dancojocaru2000** [comment]
Since iPad OS 13.4 now has mouse support, such a context menu widget should be considered for iOS as well.

**Comment by esDotDev** [comment]
fwiw, with a bit of logic to offset the context-menu depending on which quadrant off the app you are in, it becomes almost impossible to have the context menu be cut-off. If you take a look at VSCode it works like this. 
![XJEcXBJ](https://user-images.githubusercontent.com/736973/101600603-cbbf2700-39b8-11eb-9123-37ef236e72f2.gif)

I was able to code this up in about 100 lines, just using a Stack at the top of the tree and a custom Notification:
![awpMI4v](https://user-images.githubusercontent.com/736973/101600058-02487200-39b8-11eb-8bd3-50c5ba08188f.gif)

Usage is something like:
```
ContextMenuRegion(
  size: Size(100,100),
  menuBuilder: (_) => MyMenu(),
  child: MyContent(),
)
```
Certainly there are use cases for going outside the window (for very small apps, or very huge menus), but I think 98% of needs can be served with something simple like this, and the simple offset logic is actually very robust/elegant.
```
    double dx = 0, dy = 0;
    if (_mousePos.dx > _windowSize.width / 2) dx = -_menuSize.width;
    if (_mousePos.dy > _windowSize.height / 2) dy = -_menuSize.height;
```

**Comment by justinmc** [comment]
I played around with creating a Flutter-rendered context menu on Mac and getting it to work with text editing:

![desktop_tsm](https://user-images.githubusercontent.com/389558/102130742-10f2b700-3e06-11eb-951a-74195a498e93.gif)

The messy POC code is on my [desktop-context-menu](https://github.com/justinmc/flutter/tree/desktop-context-menu) branch, but I'm holding off on moving forward with it due to a few blockers:

  * It needs to share code with the Android and iOS text selection menus. These need to be refactored a bit to support this. The Android menu's refactor is about to be merged in https://github.com/flutter/flutter/pull/69428, and iOS should follow suit.  This will also allow custom menu items, which will be really important since desktop context menus often have dozens of options.
  * It's too hard to implement distinct text editing behaviors in the current code (e.g. what happens on double tap/click, when the menu should show, etc.).  Just handling some distinct behaviors in Android and iOS now is resulting in a steady stream of bugs and regressions. This needs to be refactored to support multiple platforms, which I am working on proposing in [this design doc](https://docs.google.com/document/d/1QaVIr1bbOWJGNyyNsB75sXgTDRMuxP268sLGTBNDut0/edit).
  * TextButton needs to support custom alignment so that we can make the buttons align left like on native Mac.

I didn't see @esDotDev's comment above about menu positioning when I started this, but that looks like a great way to do positioning.

**Tl;dr**: The approach of rendering a desktop context menu with Flutter looks totally doable, but we will struggle with bugs and limited features if we don't complete some refactoring first.

**Comment by justinmc** [comment]
A Mac context menu was implemented in https://github.com/flutter/flutter/pull/73882 and a generic Material one will be implemented for Mac and Linux in https://github.com/flutter/flutter/pull/74286.  I'll close this issue when it's merged.

However, these menus are super basic and have major limitations for now.  For example, for now, the only buttons available are the default cut/copy/paste/selectall from mobile, and there is no way for developers to extend these menus.  I'm tracking these next steps in https://github.com/flutter/flutter/issues/74255.

| | Light | Dark |
| -- | ----- | ----- |
| Mac | <img width="278" alt="Screen Shot 2021-01-15 at 10 37 00 AM" src="https://user-images.githubusercontent.com/389558/104765467-ce324080-571d-11eb-875b-ed37bde12db6.png"> | <img width="309" alt="Screen Shot 2021-01-15 at 10 36 45 AM" src="https://user-images.githubusercontent.com/389558/104765465-cd011380-571d-11eb-9497-c380be3f7c81.png"> |
| Windows/Linux | <img width="290" alt="Screen Shot 2021-01-20 at 9 59 16 AM" src="https://user-images.githubusercontent.com/389558/105215569-4629ad80-5b06-11eb-9649-882f67ee4560.png"> | <img width="299" alt="Screen Shot 2021-01-20 at 9 59 35 AM" src="https://user-images.githubusercontent.com/389558/105215589-4cb82500-5b06-11eb-95b2-b1d387414c78.png"> |

**Comment by aBuder** [comment]
> fwiw, with a bit of logic to offset the context-menu depending on which quadrant off the app you are in, it becomes almost impossible to have the context menu be cut-off. If you take a look at VSCode it works like this. Spotify also does a variation of this.
> 
> <img src="https://user-images.githubusercontent.com/736973/101600603-cbbf2700-39b8-11eb-9123-37ef236e72f2.gif" height=300>
> 
> I was able to code this up in about 90 lines, just using a Stack at the top of the tree and a custom Notification:
> 
> <img src="https://i.imgur.com/j622J3C.gif" width=400 />
> 
> 
> 
> Usage is something like:
> 
> ```
> 
>  MaterialApp(
> 
>       home: ContextOverlay( child: ... )
> 
> );
> 
> ...
> 
> ContextMenuRegion(
> 
>   menuBuilder: (_) => MyMenu(),
> 
>   child: SomeContent(),
> 
> )
> 
> ```
> 
> Certainly there are use cases for going outside the window (for very small apps, or very huge menus), but I think 98% of needs can be served with something simple like this, and the simple offset logic is actually very robust/elegant. 
> 
> 
> 
> Hopefully the primary use case doesn't become overly complicated in order to serve the edge case of falling outside the window chrome. Many very popular desktop apps, like Slack, Steam, Spotify, VSCode etc all use totally custom context menus, and they often look nicer for it, so it would be nice to be able to render as a widget if possible imo, otherwise it seems like styling it to match our apps could be quite hard/inflexible.

Can you share the code
